### PR TITLE
Freehand line: adjust segments before editing 

### PR DESCRIPTION
**Pull request by eatyourgreens**
- Add `revertEdits` and `cancelClipPath` actions.
- Revert any existing clipped points before starting a new edit.
- Refactor `line.cutSegment` to take the start and end of the segment as arguments: `cutSegment(startPoint, endPoint)`.
- Refactor splicing and trimming the points array to allow for drawing in either direction. The line should now reverse if you edit in the opposite direction to the original drawing.
- Refactor the freehand line state logic to use `segmentationInProgress` which will be true if either:
  - the line path (open or closed) has been cut into two segments in the middle.
  - the line is open but we are editing the ends of a closed shape.

I've also included some fixes for small bugs that came up in the review of #3559.

These changes allow the clipped path to be modified by clicking anywhere along the line path, which allows the endpoint to be adjusted before drawing a new segment.

- double-click anywhere on a line to start editing a segment.
- click anywhere to set the end of the segment. You can change your mind and click somewhere else if you make a mistake.
- click off the line to reset and start again.

https://user-images.githubusercontent.com/59547/188265662-b393fca6-02b4-4148-8f19-9a9337d079c4.mov

## Package
lib-classifier/src/plugins/drawingTools/experimental/FreehandLine

## How to Review
You can draw, and edit, freehand lines in the Classifier storybook.
http://localhost:6006/?path=/story/drawing-tools-freehand-line--complete&globals=locale:en;locales.en:English;locales.test:Test+Language

Staged project with a drawing workflow:
https://fe-project-branch.preview.zooniverse.org/projects/eatyourgreens/-project-testing-ground/classify/workflow/3370?env=staging

https://user-images.githubusercontent.com/59547/188324525-9cd78f76-d1aa-4cb1-9311-09a05c497910.mov

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## New Feature
- [ ] The PR creator has listed user actions to use when testing the new feature
- [ ] Unit tests are included for the new feature
- [ ] A storybook story has been created or updated
  - Example of SlideTutorial [component](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SlideTutorial/SlideTutorial.js) with docgen comments, and its [story](https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/docs/other-slidetutorial--default)

### CODE REVIEW

**Comment by eatyourgreens** [comment]
Draft because this hasn't been tested with a closed shapes yet.

File: packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js [hunk 1]
```diff
@@ -210,7 +240,7 @@ const FreehandLineModel = types
     trim(startIndex, endIndex) {
       const clippedPoints = [ ...self.points.slice(endIndex), ...self.points.slice(0, startIndex + 1)]
       self.clipPath = clippedPoints.map(({ x, y }) => ({ x, y }))
-      self.points.splice(0, startIndex + 1)
+      self.points.splice(0, startIndex)
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js:243]: The second argument to [`slice`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice) is an array index, but the second argument to [`splice`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice) is the number of points that you want to remove from the array. Here, we remove `startIndex` points from the beginning of the array, which moves the first point to `startIndex`.

File: packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js [hunk 2]
```diff
@@ -67,8 +67,13 @@ const FreehandLineModel = types
...
       // closes the drawing path
-      if (!self.dragPoint && self.isClosed) {
-        path.push(`Z`)
+      if (self.isClosed) {
+        if (self.dragPoint) {
+          const { x, y } = self.initialPoint
+          path.push(`L ${x},${y}`)
+        } else {
+          path.push(`Z`)
+        }
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js:76]: The [`Z` path command](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/d#closepath) closes the current subpath, which looks very weird when the line is cut into two subpaths. This works around that by drawing a line to the first point when the line has been cut.

**Comment by eatyourgreens** [comment]
Weird bug when editing near the start of a closed loop. A short segment of the original line disappeared.


https://user-images.githubusercontent.com/59547/188325916-733b8d63-39f6-45a6-9d3f-292305cd73d4.mov

**Comment by eatyourgreens** [comment]
Refactoring `line.cutSegment` as `line.cutSegment(start, end)` might have fixed that bug when editing near the start/end of a closed path.


https://user-images.githubusercontent.com/59547/188426331-fb5da2fc-2f8c-4357-b459-78abec6e79cd.mov

**Comment by eatyourgreens** [comment]
Clicking off the line is easy to do by mistake, so maybe deselecting the active mark needs [a confirmation before continuing](https://developer.mozilla.org/en-US/docs/Web/API/Window/confirm)?

**Comment by eatyourgreens** [comment]
Here's a link to [a staged branch of Etch-A-Cell](https://fe-project-branch.preview.zooniverse.org/projects/h-spiers/etch-a-cell/classify/workflow/8921) with this version of the freehand line tool.

**Comment by eatyourgreens** [comment]
Here's a screen recording, showing the latest deployed version of the code. Big changes are:
- the code respects the order in which you choose points for editing. The first double click places the draggable open circle. Second click places the filled target circle. Drawing in either direction is allowed.
- you can click multiple times to position the filled circle before you start drawing from the open circle.
- clicking anywhere off the line resets the line back to the original shape.

https://user-images.githubusercontent.com/59547/189710664-b4eb251d-a050-4e26-aa40-2ad9eaf33d55.mov

**Comment by eatyourgreens** [comment]
Here's a video with an Etch-A-Cell image, where I edit a small section of the outline of the nucleus.


https://user-images.githubusercontent.com/59547/189713986-90a7f276-e0ee-4a8f-9ad7-372b88f2de9a.mov

File: packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js [hunk 3]
```diff
@@ -201,17 +225,50 @@ const FreehandLineModel = types
...
+        const lastIndex = self.points.length - 1
+        self.points.reverse()
+        firstPoint = lastIndex - startIndex
+        lastPoint = lastIndex - endIndex
+      }
+      const deleteCount = lastPoint - firstPoint - 1
+      const clippedPoints = self.points.slice(firstPoint, lastPoint + 1)
       self.clipPath = clippedPoints.map(({ x, y }) => ({ x, y }))
-      self.points.splice(startIndex + 1, deleteCount)
+      self.points.splice(firstPoint + 1, deleteCount)
```
Comment (shaunanoordin) [inline packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js:243]: [minor] Note to self: don't worry about the _extremely unlikely_ scenario of lastPoint === firstPoint. If deleteCount is -1, the splice functions similarly to deleteCount = 0 and won't throw an exception.

**Review by shaunanoordin — APPROVED** [review]
## PR Review

Package: `lib-classifier`
Affects: experimental Freehand Line drawing tool

This PR improves the "edit" functionality of the Freehand Line tool. See #3446 for basic details.

- New behaviour: when the volunteer has _fully started_ the edit process by placing the "edit start" (by double-clicking on the mark) and "edit end" (by clicking on the mark again) points/drag handles on the FHL mark, the volunteer can _continue_ to reposition the "edit end" point _even on the currently-being-edited segment_ (the dashed lines)
- New behaviour: if the volunteer places the "edit start" point **before** the "edit end" point (in the context of the FHL's array or coordinates), then the FHL mark('s array of coordinates) will **reverse itself** _once the edit is complete._
  - From a user's perspective, the change is not noticeable, except perhaps for the positions of the "Delete" button & "continue drawing" handle changing from one end of an _open FHL Line_ to the other end.
  - From a data perspective, well, the array of coordinates is reversed, so a mark drawn as a clockwise semicircle will now be recorded as a counter-clockwise semicircle of the same shape. See Dev Notes for discussion.

_Video: the "reversing" in action._
https://user-images.githubusercontent.com/13952701/190641978-de14fb0f-6c68-40ec-a7d3-ebadcdeb0fb6.mov

All other behaviour remains the same, _notably_ the "revert/reset edit" behaviour, which reverts all edits-in-progess (at whatever stage of the edit) when the FHL mark no longer becomes the actively selected mark.

Code read was pretty complex, but I can't find any issues with the logic. I'm giving a special 👍 for distinctive naming of firstPoint/lastPoint vs startIndex/endIndex. Tests check out too, so this is looking good.

### Dev Notes

I'm all for the "reverse the array if 'edit-end' point appears before 'edit-start' point", since this is a fairly intuitive UI behaviour for the user. _Especially_ for closed shapes. 👍 

One thing to consider though is: will reversing the array affect any aggregation or machine-learning logic? i.e. to a user, a path going `[ 1, 2, 3, 4, 5]` looks visually the same as a path going `[5, 4, 3, 2, 1]`, but to a computer _that's not factoring reversibility_ might consider them two distinct paths.

### Testing

Tested on localhost with macOS + Chrome 105

Setup:
- Test URL: https://local.zooniverse.org:8080/?env=staging&project=darkeshard%2Ftest-project-2022&workflow=3596 (test project with a WF with Freehand Line drawing tool)

Tests:
- I'm able to (a) draw open FHL marks (i.e. start and end points don't join) and (b) draw closed FHL marks.
- I'm able to _start_ the edit process by (1) double clicking to place the "edit start" point, and (2) clicking to place the "edit end" point
  - I can continue placing the "edit end" by clicking on the FHL mark, _including on the currently-being-edited segment_ (dashed line)
  - If I place the "edit end" _before_ the "edit start", the FHL's coordinate array reverses. (Noticeable with open FHL shapes) 
  - I can cancel out of the edit process at any time by deselecting the mark (e.g. by changing tools or selecting another mark), and all edits will be reverted, _including any "reversing" of arrays._
- I'm able to complete the edit by (3) dragging the "edit start" drag handle to the "edit end" point.

### Status

LGTM. 👍 It might be worth asking the team if the reversibility of the FHL coord array is worth noting, but as far as I'm concerned, this PR is good to go.

**Comment by eatyourgreens** [comment]
>One thing to consider though is: will reversing the array affect any aggregation or machine-learning logic? i.e. to a user, a path going [ 1, 2, 3, 4, 5] looks visually the same as a path going [5, 4, 3, 2, 1], but to a computer that's not factoring reversibility might consider them two distinct paths.

The live freehand drawing projects allow you to pick any starting point, and draw in either direction, so I'd guess that this is already a solved problem for data analysis.

I'm more worried about slow performance of these tools (plus overly large classification payloads.) Watch the lag here between clicking with the mouse and seeing the SVG line update on a partially drawn Etch-A-Cell cell nucleus. Array operations are expensive when the array has hundreds or thousands of points.

There were open issues about poor performance of the freehand drawing tools in PFE but the current monorepo code hasn't yet addressed those problems.


https://user-images.githubusercontent.com/59547/190652929-4de050cb-bd69-4d46-b790-b64d65b9cedd.mov

**Pull request merged**